crates/mysten-metrics/src/monitored_mpsc.rs

```diff
@@ -0,0 +1,855 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Provides wrappers to tokio mpsc channels, with metrics on total items sent, received and inflight.
+
+use std::task::{Context, Poll};
+
+use futures::{Future, TryFutureExt as _};
+use prometheus::IntGauge;
+use tap::Tap;
+use tokio::sync::mpsc::{
+    self,
+    error::{SendError, TryRecvError, TrySendError},
+};
+
+use crate::get_metrics;
+
+/// Wraps [`mpsc::Sender`] with gauges counting the sent and inflight items.
+#[derive(Debug)]
+pub struct Sender<T> {
+    inner: mpsc::Sender<T>,
+    inflight: Option<IntGauge>,
+    sent: Option<IntGauge>,
+}
+
+impl<T> Sender<T> {
+    /// Sends a value, waiting until there is capacity.
+    /// Increments the gauge in case of a successful `send`.
+    pub async fn send(&self, value: T) -> Result<(), SendError<T>> {
+        self.inner
+            .send(value)
+            .inspect_ok(|_| {
+                if let Some(inflight) = &self.inflight {
+                    inflight.inc();
+                }
+                if let Some(sent) = &self.sent {
+                    sent.inc();
+                }
+            })
+            .await
+    }
+
+    /// Completes when the receiver has dropped.
+    pub async fn closed(&self) {
+        self.inner.closed().await
+    }
+
+    /// Attempts to immediately send a message on this `Sender`
+    /// Increments the gauge in case of a successful `try_send`.
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        self.inner
+            .try_send(message)
+            // remove this unsightly hack once https://github.com/rust-lang/rust/issues/91345 is resolved
```

Comment:
looks like the issue is resolved

Reply:
Yes but unfortunately we need to upgrade Rust to use the inspect API.